scrapy_zyte_api/_downloader_middleware.py

```diff
@@ -23,3 +36,22 @@ def process_request(self, request, spider):
             request.meta["download_slot"] = slot_id
         _, slot = downloader._get_slot(request, spider)
         slot.delay = 0
+
+        if self._max_requests_reached(downloader):
+            self._crawler.engine.close_spider(spider, "closespider_max_zapi_requests")
+            raise IgnoreRequest("Reached max Zyte API requests")
```

Comment:
As, I think, this code will be executed several times, I wonder how will it look in the log, hopefully not too ugly.

Reply:
There should just be a few of them since the spider is immediately closed.

Reply:
As long as the message reads “this request is skipped because the max has been reached” rather than just “the max has been reached”, I think it is OK for this to appear several times.

Reply:
Message updated in https://github.com/scrapy-plugins/scrapy-zyte-api/pull/114/commits/660cdb2ab5cf04c30c058c6f471f1ad8044e5876